src/main/java/duke/tasks/TaskList.java

```diff
@@ -0,0 +1,173 @@
+ * to invalid task inputs or operations.
+ * </p>
+ */
+public class TaskList {
+
+    /**
+     * The list of tasks being managed by the {@code TaskManager}.
+     */
+    private List<Task> tasks;
+
+    /**
+     * Constructs a new {@code TaskManager} with an empty task list.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    /**
+     * Retrieves the list of tasks managed by this {@code TaskManager}.
+     *
+     * @return an {@code ArrayList} of tasks.
+     */
+    public List<Task> getTasks() {
+        return tasks;
+    }
+
+    /**
+     * Retrieves the task at the specified index in the task list.
+     *
+     * @param index the index of the task to retrieve.
+     * @return the {@code Task} at the specified index.
+     * @throws IndexOutOfBoundsException if the index is out of the range of the task list.
+     */
+    public Task getTask(int index) {
+        if (index >= 0 && index < tasks.size()) {
+            return tasks.get(index);
+        }
+        throw new IndexOutOfBoundsException("Invalid task index.");
+    }
+
+    /**
+     * Retrieves the last task in the task list.
+     *
+     * @return the last {@code Task} in the task list.
+     * @throws IllegalStateException if the task list is empty.
+     */
+    public Task getLastTask() {
+        if (tasks.isEmpty()) {
+            throw new IllegalStateException("No tasks available.");
+        }
+        return tasks.get(tasks.size() - 1);
+    }
+
+    public List<Task> getTasksOccurring(LocalDateTime dateTime) {
+        List<Task> tasksOccurringOnDateTime = new ArrayList<>();
+
+        for (Task task : this.tasks) {
+            if (task.occurring(dateTime)) {
+                tasksOccurringOnDateTime.add(task);
+            }
+        }
+
+        return tasksOccurringOnDateTime;
+    }
+
+    public void setTasks(List<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Deletes the task at the specified index from the task list.
+     *
+     * @param index the index of the task to delete.
+     * @return the {@code Task} that was removed.
+     * @throws IndexOutOfBoundsException if the index is out of the range of the task list.
+     */
+    public Task deleteTask(int index) {
+        if (index >= 0 && index < tasks.size()) {
+            return tasks.remove(index);
+        }
+        throw new IndexOutOfBoundsException("Invalid task index.");
+    }
+
+    /**
+     * Adds new tasks to the task list based on the user's inputs.
+     * Each input is parsed to determine the type of task (ToDo, Deadline, or Event).
+     *
+     * @param userInputs the raw input strings from the user.
+     * @throws UnknownMessageException if the task type is unknown or unrecognized.
+     * @throws InvalidTodoDescriptionException if the task is a ToDo but lacks a description.
+     */
+    public String addTask(String... userInputs) throws UnknownMessageException, InvalidTodoDescriptionException {
```

Comment:
I really like the different types of exceptions you have created, which are self-descriptive